rust/worker/src/segment/types.rs

```diff
@@ -136,131 +140,274 @@ pub struct MaterializedLogRecord<'referred_data> {
+            metadata_to_be_deleted: None,
+            final_document_at_log_index: None,
+            final_embedding_at_log_index: None,
         }
-        match self.final_document {
-            Some(doc) => Some(doc.to_string()),
-            None => match self.data_record.as_ref() {
-                Some(data_record) => data_record.document.map(|doc| doc.to_string()),
-                None => None,
+    }
+
+    fn from_log_record(
+        offset_id: u32,
+        log_index: usize,
+        log_record: &LogRecord,
+    ) -> Result<Self, LogMaterializerError> {
+        let final_document_at_log_index = if log_record.record.document.is_some() {
+            Some(log_index)
+        } else {
+            None
+        };
+
+        let final_embedding_at_log_index = if log_record.record.embedding.is_some() {
+            Some(log_index)
+        } else {
+            return Err(LogMaterializerError::EmbeddingMaterialization);
+        };
+
+        let merged_metadata;
+        let deleted_metadata;
+        match &log_record.record.metadata {
+            Some(metadata) => match materialize_update_metadata(metadata) {
+                Ok(m) => {
+                    merged_metadata = Some(m.0);
+                    deleted_metadata = Some(m.1);
+                }
+                Err(e) => {
+                    return Err(LogMaterializerError::MetadataMaterialization(e));
+                }
             },
-        }
+            None => {
+                merged_metadata = None;
+                deleted_metadata = None;
+            }
+        };
+
+        Ok(Self {
+            offset_id_exists_in_segment: false,
+            offset_id,
+            user_id_at_log_index: Some(log_index),
+            final_operation: MaterializedLogOperation::AddNew,
+            metadata_to_be_merged: merged_metadata,
+            metadata_to_be_deleted: deleted_metadata,
+            final_document_at_log_index,
+            final_embedding_at_log_index,
+        })
     }
+}
 
-    pub(crate) fn merged_document_ref(&self) -> Option<&str> {
-        if self.final_operation == MaterializedLogOperation::OverwriteExisting
-            || self.final_operation == MaterializedLogOperation::AddNew
-        {
-            return match self.final_document {
-                Some(doc) => Some(doc),
-                None => None,
-            };
-        }
-        match self.final_document {
-            Some(doc) => Some(doc),
-            None => match self.data_record.as_ref() {
-                Some(data_record) => match data_record.document {
-                    Some(doc) => Some(doc),
-                    None => None,
-                },
+/// Obtained from a `MaterializeLogsResult`. Provides a borrowed view of a single materialized log record.
+/// You will probably need to call `.hydrate()` on this struct for most use cases, although you can view the offset ID and operation without hydrating.
+pub struct BorrowedMaterializedLogRecord<'log_data> {
```

Comment:
can you leave a comment as to why this intermediary struct is needed instead of MaterializedLogRecord and its hydrated counterpart.